File: cpp/refactored.cpp
#include <iostream>

using namespace std;
// Low-level class (can remain unchanged).
class Wallet {
private:
    float money;

public:
    Wallet(float initialMoney) : money(initialMoney) {}

    float getMoney() const {
        return money;
    }

    bool hasSufficientFunds(float amount) const {
        return money >= amount;
    }

    void debit(float amount) {
        if (hasSufficientFunds(amount)) {
            money -= amount;
        }
    }
};

// Mid-level class now handles its own internal logic.
class Customer {
private:
    Wallet wallet;

public:
    Customer(float money) : wallet(money) {}

    // REFACTORED: The Customer now provides a high-level method to make a payment.
    // It does NOT expose its internal wallet. It handles the details itself,
    // thereby hiding its implementation.
    void makePayment(float amount) {
        cout << "Customer is making a payment of $" << amount << endl;
        if (wallet.hasSufficientFunds(amount)) {
            wallet.debit(amount);
        } else {
            cout << "Customer has insufficient funds." << endl;
        }
    }

    float checkBalance() const {
        return wallet.getMoney();
    }
};

// High-level class is now much simpler and decoupled.
class Store {
public:
    // REFACTORED: The Store talks ONLY to its "immediate friend," the Customer.
    // It tells the customer what it wants (to make a payment) but not HOW to do it.
    // It has no knowledge of a "Wallet".
    void processPayment(Customer& customer, float amount) {
        cout << "Store is requesting payment from customer..." << endl;

        // The Store delegates the payment task to the Customer.
        customer.makePayment(amount);
        
        float remaining = customer.checkBalance();
        cout << "Payment processed. Customer now has $" << remaining << " left." << endl;
    }
};

int main() {
    Customer customer(100.0f);
    Store store;
    store.processPayment(customer, 25.0f);
    return 0;
}

File: pyhton/refactored.py
# refactored.py
from copy import deepcopy

# --- Functions that know about the "Customer" data structure ---

"""A helper function that knows how to get a customer's balance."""
def get_customer_balance(customer: dict) -> float:
    return customer['wallet']['money']


"""
Handles the logic for a customer payment. This is the "friend" function
that is allowed to know the customer's internal structure.
It returns a NEW, updated customer state to promote immutability.
"""
def customer_pays(customer: dict, amount: float) -> dict:

    print(f"{customer['name']} is making a payment of ${amount:.2f}")

    if get_customer_balance(customer) >= amount:
        # Create a deep copy to avoid changing the original data (immutability).
        new_customer = deepcopy(customer)
        new_customer['wallet']['money'] -= amount
        return new_customer

    # If not enough money, return the original state unchanged.
    print(f"{customer['name']} has insufficient funds.")
    return customer

"""
REFACTORED: This function is now decoupled from the Customer's data structure.
It talks ONLY to its "immediate friend" functions (`customer_pays`, `get_customer_balance`).
It delegates the tasks instead of doing them itself. It has no knowledge
of a 'wallet' or 'money'.
"""
# --- Function that knows about the "Store" logic ---

def process_payment(store_name: str, customer: dict, amount:float) -> dict:

    print(f"Store '{store_name}' is requesting payment...")
    
    # Delegate the payment task to the appropriate helper function.
    updated_customer = customer_pays(customer, amount)

    remaining = get_customer_balance(updated_customer)
    print(f"Payment processed. {updated_customer['name']} has ${remaining:.2f} left.")

    return updated_customer

# --- Main execution ---
if __name__ == "__main__":
    # Define our initial, immutable data state.
    customer_data = {
        'name': 'Bob',
        'wallet': {'money': 100.0}
    }
    store_name = "The Corner Shop"

    # The function receives the old state and returns the new state.
    final_customer_state = process_payment(store_name, customer_data, 25.0)

    print(f"\nOriginal state was not changed: {customer_data}")
    print(f"Final state: {final_customer_state}")
